8155 external metadata validation

landreev opened the pull request:
**What this PR does / why we need it**:

The code in the PR has been reviewed, and QA-ed, over the months (almost a year) it's been in use in prod. 
The reason we need it, I will no longer have to build custom release patches.

**Which issue(s) this PR closes**:

Closes #8155

**Special notes for your reviewer**:

see the description

**Suggestions on how to test this**:

see the description 

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

**Is there a release notes update needed for this change?**:

**Additional documentation**:

pdurbin commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -2409,3 +2409,55 @@ setting indicates embargoes are not supported. A value of -1 allows embargoes of
+:DataverseMetadataValidatorScript
++++++++++++++++++++++++++++++++++
+
+An optional external script that validates Dataverse collection metadata as it's being updated or published. The script provided should be an executable that takes a single command line argument, the name of the file containing the metadata exported in the native json format. The script should exit with a non-zero error code if the validation fails. If that happens, a failure message (customizabl ...
```
```suggestion
An optional external script that validates Dataverse collection metadata as it's being updated or published. The script provided should be an executable that takes a single command line argument, the full path to the Dataverse collection validation script. The script should exit with a non-zero error code if the validation fails. If that happens, a failure message (customizable in the next two settings below, `:DataverseMetadataValidationFailureMsg` and `:DataverseMetadataUpdateValidationFailureMsg`) will be shown to the user.
```

landreev replied:
No, that would be completely incorrect - sorry.

landreev replied:
(fixed the missing colon though)

pdurbin replied:
Ah, I thought I had found a copy/paste error. I get it now. There's a lot packed into a single sentence. It's fine to leave it alone but we could also consider separated describing:

- what the setting points to (the script)
- how the script operates (by taking a JSON file as an argument)

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/util/SystemConfig.java:
```diff
@@ -1091,7 +1091,46 @@ public String getDataCiteRestApiUrlString() {
+    }
+    
+    public String getDataverseValidationFailureMsg() {
+        String defaultMessage = "This dataverse collection cannot be published because it has failed an external metadata validation test.";
+        return settingsService.getValueForKey(SettingsServiceBean.Key.DataverseMetadataValidationFailureMsg, defaultMessage);
+    }
+    
+    public String getDataverseUpdateValidationFailureMsg() {
+        String defaultMessage = "This dataverse collection cannot be updated because it has failed an external metadata validation test.";
+        return settingsService.getValueForKey(SettingsServiceBean.Key.DataverseMetadataUpdateValidationFailureMsg, defaultMessage);
+    }
+    
+    public String getDatasetValidationFailureMsg() {
+        String defaultMessage = "This dataset cannot be published because it has failed an external metadata validation test.";
```
These English strings should probably go in Bundle.properties since people might be running Dataverse in a non-English language.

landreev replied:
... but since we are providing the settings for customizing these messages - sounds kinda redundant, no?

pdurbin replied:
Except that you might have an installation with three languages configured. The speakers of each language should see the error in their native tongue, I assume.

I mean, maybe it'll be years before this scenario happens (multiple langs and spam check configured). We could address it later.

pdurbin commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -2409,3 +2409,55 @@ setting indicates embargoes are not supported. A value of -1 allows embargoes of
+``curl -X PUT -d /usr/local/bin/dv_validator.sh http://localhost:8080/api/admin/settings/:DataverseMetadataValidatorScript``
+
+
+:DataverseMetadataValidationFailureMsg
```
:DataverseMetadataValidationPublishFailureMsg might be more clear as a setting name.

Otherwise, the only difference between this setting and the one below is the presence or absence of the word "update".

landreev replied:
it's pure legacy. we thought it would only need to be done on publish, initially, just like for datasets. Then remembered that there is no "draft" published dataverses - so somebody could publish an innocuous dataverse, then add spam text to the description...

pdurbin commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -2409,3 +2409,55 @@ setting indicates embargoes are not supported. A value of -1 allows embargoes of
+:DatasetMetadataValidatorScript
++++++++++++++++++++++++++++++++
+
+An optional external script that validates dataset metadata during publishing. The script provided should be an executable that takes a single command line argument, the name of the file containing the metadata exported in the native json format. The script should exit with a non-zero error code if the validation fails. If that happens, the dataset is left unpublished, and a failure message (cust ...
```
```suggestion
An optional external script that validates dataset metadata during publishing. The script provided should be an executable that takes a single command line argument, the full path to the dataset validation script. The script should exit with a non-zero error code if the validation fails. If that happens, the dataset is left unpublished, and a failure message (customizable in the next setting below, `:DatasetMetadataValidationFailureMsg`) will be shown to the user.
```

pdurbin commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -2409,3 +2409,55 @@ setting indicates embargoes are not supported. A value of -1 allows embargoes of
 
 ``curl -X PUT -d 24 http://localhost:8080/api/admin/settings/:MaxEmbargoDurationInMonths``
+
+:DataverseMetadataValidatorScript
```
Just a thought, you could use the word "collection" for the dataverse settings, such as `:CollectionMetadataValidatorScript`.

landreev replied:
These existing settings have been in use for a year in prod. I'm not messing with them. Legacy.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/UpdateDataverseCommand.java:
```diff
@@ -57,6 +61,22 @@ public UpdateDataverseCommand(Dataverse editedDv, List<DatasetFieldType> facetLi
 	
 	@Override
 	public Dataverse execute(CommandContext ctxt) throws CommandException {
+            logger.info("Entering update dataverse command");
```
This should probably be reduced to logger.fine.

pdurbin reviewed (commented):
Overall this looks great but I left some comments, some of which should probably be addressed.

pdurbin commented on doc/sphinx-guides/source/installation/config.rst:
```diff
@@ -2409,3 +2409,55 @@ setting indicates embargoes are not supported. A value of -1 allows embargoes of
+
+
+:DataverseMetadataPublishValidationFailureMsg
+++++++++++++++++++++++++++++++++++++++
```
`make html` is failing with this:

installation/config.rst:2424:Title underline too short.

landreev commented:
OK, fixed the broken underline, and added clarifying sentences, to the script setting sections. 
Should be ready now, correct - since we agreed not to tinker with the language customization in this PR?

pdurbin reviewed (approved):
Looks good. Yes, we can wait on moving English to the bundle until we have an installation that wants to use this code AND has multiple languages configured.
